File: solutions/csharp/circular-buffer/1/CircularBuffer.cs
public class CircularBuffer<T>
{
    private readonly T?[] _buffer;
    private int _bufferLength;
    private int _readIndex;
    private int _writeIndex;

    public CircularBuffer(int capacity)
    {
        _buffer = new T?[capacity];
        _writeIndex = GetRandomIndex();
        _readIndex = _writeIndex;
        _bufferLength = 0;
    }

    public T Read()
    {
        if(IsEmpty())
            throw new InvalidOperationException("Can not read from an empty buffer.");
        T value = _buffer[_readIndex] ?? throw new NullReferenceException("Buffer read value should not be null.");
        _readIndex = GetNextCircularIndex(_readIndex);
        _bufferLength--;
        return value;
    }

    public void Write(T value)
    {
        if(IsFull())
            throw new InvalidOperationException("Can not write to a full buffer.");
        _buffer[_writeIndex] = value;
        _writeIndex = GetNextCircularIndex(_writeIndex);
        _bufferLength++;
    }

    public void Overwrite(T value)
    {
        if (IsFull())
        {
            _buffer[_readIndex] = value;
            _readIndex = GetNextCircularIndex(_readIndex);
        }
        else
        {
            Write(value);
        }
    }

    public void Clear()
    {
        for(int i = 0; i < _bufferLength; i++)
            _buffer[i] = default;
        _writeIndex = GetRandomIndex();
        _readIndex = _writeIndex;
        _bufferLength = 0;
    }

    private int GetRandomIndex() => (new Random()).Next(0, _buffer.Length);

    private int GetNextCircularIndex(int index)
    {
        index += 1;
        if (index >= _buffer.Length)
            return index % _buffer.Length;
        return index;
    }

    private bool IsFull() => _bufferLength == _buffer.Length;

    private bool IsEmpty() => _bufferLength == 0;
}

File: solutions/cpp/circular-buffer/1/circular_buffer.cpp
#include "circular_buffer.h"
#include <stdexcept>
#include <string>

namespace circular_buffer {

template <typename T>
circular_buffer<T>::circular_buffer(const size_t capacity) : buffer(capacity) {
  buffer.reserve(capacity);
}

template <typename T> T circular_buffer<T>::read() {
  if (curr_size == 0) {
    throw std::domain_error("Buffer is empty!");
  }
  size_t prev_read_ptr = read_ptr;
  increment_ptr(read_ptr);
  --curr_size;
  return buffer[prev_read_ptr];
}

template <typename T> void circular_buffer<T>::write(T val) {
  if (curr_size == buffer.size()) {
    throw std::domain_error("Buffer is full!");
  }
  buffer[write_ptr] = val;
  increment_ptr(write_ptr);
  ++curr_size;
}

template <typename T> void circular_buffer<T>::overwrite(T val) {
  if (curr_size < buffer.size()) {
    write(val);
  } else {
    buffer[read_ptr] = val;
    increment_ptr(read_ptr);
  }
}
template <typename T> void circular_buffer<T>::clear() {
  write_ptr = 0;
  read_ptr = 0;
  curr_size = 0;
}

template <typename T> void circular_buffer<T>::increment_ptr(size_t &ptr) {
  ++ptr;
  if (ptr == buffer.size()) {
    ptr = 0;
  }
}

} // namespace circular_buffer
template class circular_buffer::circular_buffer<int>;
template class circular_buffer::circular_buffer<std::string>;
